File: tests/test_project_tree.py
import project_tree
from app import app


def test_project_tree_returns_bounded_structured_nodes():
    client = app.test_client()
    response = client.get("/api/project-tree")
    assert response.status_code == 200
    data = response.get_json()
    assert isinstance(data["nodes"], list)
    if data["nodes"]:
        node = data["nodes"][0]
        assert {"name", "path", "kind", "icon"} <= set(node)

File: tests/test_project_tree.js
const assert = require("node:assert/strict");
const { BrowserShim } = require("./browser_dom");

(async () => {
  const shim = new BrowserShim('<button id="project-tree-btn" type="button"></button>');
  const loaded = shim.load(["static/core_api.js", "static/project_tree.js"], {
    fetch: async () => ({ ok: true, json: async () => ({ nodes: [] }) }),
  });
  loaded.window.AliceDispatcher = {
    request: async () => ({ ok: true, json: async () => ({ nodes: [] }) }),
  };

  assert.equal(typeof loaded.window.ProjectTree.load, "function");
  const button = loaded.document.getElementById("project-tree-btn");
  assert.ok(button, "project tree trigger should exist");

  const loader = loaded.window.ProjectTree.load;
  await loader();

  loaded.window.AliceDispatcher.request = async () => ({ ok: false, status: 503 });
  await assert.rejects(loader, /HTTP 503/);

  console.log("project tree regression checks passed");
})().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
